ostd/src/arch/x86/timer/apic.rs

```diff
@@ -1,76 +1,111 @@
+pub(super) fn init_bsp() -> IrqLine {
     if is_tsc_deadline_mode_supported() {
-        info!("[Timer]: Enable APIC TSC deadline mode.");
-        init_tsc_mode()
+        init_deadline_mode_state();
+
+        let timer_irq = IrqLine::alloc().unwrap();
+        init_tsc_mode(&timer_irq);
+        timer_irq
     } else {
-        info!("[Timer]: Enable APIC periodic mode.");
-        init_periodic_mode()
+        init_periodic_mode_state();
+
+        let timer_irq = IrqLine::alloc().unwrap();
+        init_periodic_mode(&timer_irq);
+        timer_irq
     }
 }
 
-pub(super) static APIC_TIMER_CALLBACK: Once<Arc<dyn Fn() + Sync + Send>> = Once::new();
-
 /// Determines if the current system supports tsc_deadline mode APIC timer
 fn is_tsc_deadline_mode_supported() -> bool {
+    use x86::cpuid::cpuid;
+
     const TSC_DEADLINE_MODE_SUPPORT: u32 = 1 << 24;
     let cpuid = cpuid!(1);
     (cpuid.ecx & TSC_DEADLINE_MODE_SUPPORT) > 0
 }
 
-fn init_tsc_mode() -> IrqLine {
-    let timer_irq = IrqLine::alloc().unwrap();
+pub(super) fn init_ap(timer_irq: &IrqLine) {
+    match STATE.get() {
+        Some(State::DeadlineMode { .. }) => {
+            init_tsc_mode(timer_irq);
+        }
+        Some(State::PeriodicMode { .. }) => {
+            init_periodic_mode(timer_irq);
+        }
+        None => {
+            panic!("APIC timer state not initialized");
+        }
+    }
+}
+
+fn init_tsc_mode(timer_irq: &IrqLine) {
     // Enable tsc deadline mode
     apic::with_borrow(|apic| {
         apic.set_lvt_timer(timer_irq.num() as u64 | (1 << 18));
     });
-    let tsc_step = TSC_FREQ.load(Ordering::Relaxed) / TIMER_FREQ;
 
-    let callback = move || unsafe {
-        let tsc_value = _rdtsc();
-        let next_tsc_value = tsc_step + tsc_value;
-        wrmsr(IA32_TSC_DEADLINE, next_tsc_value);
+    apic_tsc_deadline_mode_callback();
+}
+
+pub(super) fn apic_tsc_deadline_mode_callback() {
+    use x86::msr::{wrmsr, IA32_TSC_DEADLINE};
+
+    let tsc_value = unsafe { _rdtsc() };
+    let Some(State::DeadlineMode { init_count }) = STATE.get() else {
+        panic!("APIC timer state not initialized as deadline mode");
     };
+    let next_tsc_value = init_count + tsc_value;
+    unsafe { wrmsr(IA32_TSC_DEADLINE, next_tsc_value) };
+}
 
-    callback.call(());
-    APIC_TIMER_CALLBACK.call_once(|| Arc::new(callback));
+fn init_periodic_mode(timer_irq: &IrqLine) {
+    apic::with_borrow(|apic| {
+        let Some(State::PeriodicMode { tsc_interval }) = STATE.get() else {
+            panic!("APIC timer state not initialized as periodic mode");
+        };
+        apic.set_timer_init_count(*tsc_interval);
+        apic.set_lvt_timer(timer_irq.num() as u64 | (1 << 17));
+        apic.set_timer_div_config(DivideConfig::Divide64);
+    });
+}
 
-    timer_irq
+pub(super) static STATE: spin::Once<State> = spin::Once::new();
+
+pub(super) enum State {
+    DeadlineMode { init_count: u64 },
+    PeriodicMode { tsc_interval: u64 },
```

Comment:
I'm afraid that the names are wrong..
```suggestion
    DeadlineMode { tsc_interval: u64 },
    PeriodicMode { init_count: u64 },
```